src/source/tile.js

```diff
@@ -206,78 +181,58 @@ class Tile {
         }
 
         this.collisionBoxArray = null;
-        this.collisionTile = null;
         this.featureIndex = null;
         this.state = 'unloaded';
+        this.crossTileSymbolIndex = null;
     }
 
-    redoPlacement(source: any) {
-        if (source.type !== 'vector' && source.type !== 'geojson') {
-            return;
-        }
-        if (this.state !== 'loaded') {
-            this.redoWhenDone = true;
-            return;
-        }
-        if (!this.collisionTile) { // empty tile
-            return;
+    added() {
+        for (const id in this.buckets) {
+            const bucket = this.buckets[id];
+            if (bucket instanceof SymbolBucket) {
+                this.crossTileSymbolIndex.addTileLayer(id, this.coord, this.sourceMaxZoom, bucket.symbolInstances);
+            }
         }
+    }
 
-        const cameraToTileDistance = source.map.transform.cameraToTileDistance(this);
-        if (this.angle === source.map.transform.angle &&
-            this.pitch === source.map.transform.pitch &&
-            this.showCollisionBoxes === source.map.showCollisionBoxes) {
-            if (this.cameraToTileDistance === cameraToTileDistance &&
-                this.cameraToCenterDistance === source.map.transform.cameraToCenterDistance) {
-                return;
-            } else if (this.pitch < 25) {
-                // At low pitch tile distance doesn't affect placement very
-                // much, so we skip the cost of redoPlacement
-                // However, we might as well store the latest value of
-                // cameraToTileDistance and cameraToCenterDistance in case a redoPlacement request
-                // is already queued.
-                this.cameraToTileDistance = cameraToTileDistance;
-                this.cameraToCenterDistance = source.map.transform.cameraToCenterDistance;
-                return;
+    removed() {
+        for (const id in this.buckets) {
+            const bucket = this.buckets[id];
+            if (bucket instanceof SymbolBucket) {
+                this.crossTileSymbolIndex.removeTileLayer(id, this.coord, this.sourceMaxZoom);
             }
         }
+    }
 
-        this.angle = source.map.transform.angle;
-        this.pitch = source.map.transform.pitch;
-        this.cameraToCenterDistance = source.map.transform.cameraToCenterDistance;
-        this.cameraToTileDistance = cameraToTileDistance;
-        this.showCollisionBoxes = source.map.showCollisionBoxes;
-        this.placementSource = source;
+    placeLayer(showCollisionBoxes: boolean, collisionIndex: CollisionIndex, layer: any, posMatrix: Float32Array) {
+        const bucket = this.getBucket(layer);
+        const collisionBoxArray = this.collisionBoxArray;
 
-        this.state = 'reloading';
-        this.placementThrottler.invoke();
+        if (bucket && bucket instanceof SymbolBucket && collisionBoxArray) {
+            collisionIndex.setMatrix(posMatrix);
```

Comment:
Maybe we should remove `collisionIndex.setMatrix` and instead pass the matrix as an argument to the relevant method.

Reply:
Yeah, sounds good.

Reply:
Done.